icu4c/source/test/intltest/usetheaderonlytest.cpp

```diff
@@ -0,0 +1,219 @@
+        assertEquals(WHERE, u'a', *cpIter++);
+        assertEquals(WHERE, u'b', *cpIter);
+        assertTrue(WHERE, cpIter != cpLimit);
+        CodePointRange::iterator cpIter2(u'b');  // public constructor
+        assertTrue(WHERE, cpIter == cpIter2);
+        assertEquals(WHERE, u'c', *++cpIter);
+        assertTrue(WHERE, cpIter != cpIter2);
+        assertTrue(WHERE, ++cpIter == cpLimit);
+    }
+    ++iter;
+    auto iter2(iter);  // copy constructor
+    assertEquals(WHERE, u'ç', (*iter2).rangeStart);
+    assertEquals(WHERE, u'ç', (*iter2).rangeEnd);
+    assertEquals(WHERE, 1, (*iter2).size());
+    assertEquals(WHERE, u'ç', (*iter2++).rangeStart);
+    assertEquals(WHERE, u'カ', (*iter2).rangeStart);
+    assertTrue(WHERE, ++iter2 != limit);
+    auto iter3(iter2++);
+    assertEquals(WHERE, U'🚴', (*iter3).rangeStart);
+    assertTrue(WHERE, iter2 == limit);
+
+    {
+        CodePointRange cpRange(u'h', u'k');  // public constructor
+        // FYI: currently no operator==
+        assertEquals(WHERE, u'h', cpRange.rangeStart);
+        assertEquals(WHERE, u'k', cpRange.rangeEnd);
+        assertEquals(WHERE, 4, cpRange.size());
+        assertEquals(WHERE, u'i', *++(cpRange.begin()));
+    }
+}
+
+void USetHeaderOnlyTest::TestUSetStringIterator() {
+    IcuTestErrorCode errorCode(*this, "TestUSetStringIterator");
+    using U_HEADER_NESTED_NAMESPACE::USetStrings;
+    LocalUSetPointer uset(uset_openPattern(u"[abcçカ🚴{}{abc}{de}]", -1, errorCode));
+    std::u16string result;
+    for (auto s : USetStrings(uset.getAlias())) {
+        // Commented-out sample code for pasting into the API docs.
+        // Needs U_SHOW_CPLUSPLUS_API=1 for UnicodeString.
+        // UnicodeString us(s);
+        // std::string u8;
+        // printf("uset.string length %ld \"%s\"\n", (long)s.length(), us.toUTF8String(u8).c_str());
+        result.append(u" \"").append(s).append(u"\"");
+    }
+    assertEquals(WHERE, uR"( "" "abc" "de")", result);
+
+    USetStrings range1(uset.getAlias());
+    auto range2(range1);  // copy constructor
+    auto iter = range1.begin();
+    auto limit = range2.end();
+    // operator* with pre- and post-increment
+    assertEquals(WHERE, u"", *iter);
+    assertEquals(WHERE, u"abc", *++iter);
+    auto iter2(iter);  // copy constructor
+    assertEquals(WHERE, u"abc", *iter2++);
+    assertTrue(WHERE, iter2 != limit);
+    auto iter3(iter2++);
+    assertEquals(WHERE, u"de", *iter3);
+    assertTrue(WHERE, iter2 == limit);
+}
+
+void USetHeaderOnlyTest::TestUSetElementIterator() {
+    IcuTestErrorCode errorCode(*this, "TestUSetElementIterator");
+    using U_HEADER_NESTED_NAMESPACE::USetElements;
+    LocalUSetPointer uset(uset_openPattern(u"[abcçカ🚴{}{abc}{de}]", -1, errorCode));
+    std::u16string result;
+    for (auto el : USetElements(uset.getAlias())) {
+        // Commented-out sample code for pasting into the API docs.
+        // Needs U_SHOW_CPLUSPLUS_API=1 for UnicodeString.
```

Comment:
I am a bit confused.

If we are giving sample code for an API that does not require the C++ API, shouldn’t we use a sample that doesn’t use it ?

And if we are giving sample code that does require this API, shouldn’t we test it in the other test where we do require the C++ API ?

Reply:
1. I am a bit confused why we are having an extended discussion about pre-existing, commented-out sample code. I am just moving it from one file to another, and adding one more commented-out line to make it work when not commented out.
2. The sample code is using UnicodeString for conversion to UTF-8. I concede that this does not work for someone using U_SHOW_CPLUSPLUS_API=0, but it makes for nice, succinct sample code.
3. I don't know how to output a UTF-16 string in portable C++ when not using ICU. Let me know if you have suggestions.
4. I suppose I could change this sample code to calling ICU C (ucnv.h) APIs, but then there would be more code dealing with conversion than showing how to use the iteration API.
5. I suppose I could just omit the print statement and write a comment for what the UTF-16 string will contain in each iteration.
6. Can we please deal with the commented-out sample code in a follow-up pull request?

Reply:
Re 6, sure.

Reply:
I just took a look:
- https://en.cppreference.com/w/cpp/locale/codecvt_utf8_utf16 -- zombie: added in C++11, deprecated in C++17, removed in C++26
- https://en.cppreference.com/w/cpp/locale/codecvt -- zombie:
  - std::codecvt<char16_t, char, std::mbstate_t> was added in C++11 but deprecated in C++20
  - std::codecvt<char16_t, char8_t, std::mbstate_t> was apparently born deprecated in C++20

Reply:
PS: I guess the most obvious thing that *should just work*™ would be `cout << s` where s is a std::u16string -- but I am pretty sure that that does not work.

Reply:
Fixed in
- https://github.com/unicode-org/icu/pull/3316

Reply:
> 1. I am a bit confused why we are having an extended discussion about pre-existing, commented-out sample code.

Because the context of that code was changed.